### ISSUE TITLE
Support for optimizers that require a closure

### ISSUE THREAD

**Issue by tpietruszka** [dnouri/skorch#207]
Right now, the [toy examples](https://nbviewer.jupyter.org/github/dnouri/skorch/blob/master/notebooks/Basic_Usage.ipynb) break if the `torch.optim.LBFGS` optimizer is used (likely also some others, e.g. conjugate gradients).

As per https://pytorch.org/docs/stable/optim.html#optimizer-step-closure it would seem that the closure way of calling the optimizer is preferred (and certainly required by those optimizers)

Right now, for my purposes, I've subclassed the `skorch.NeuralNetRegressor` class, overriding the `train_step` method with the following code. It seems to work, I'm just not sure if it interferes with some other skorch machinery. 
```
    def train_step(self, Xi, yi, **fit_params):
        y_pred = None
        loss = None
        def train_closure():
            nonlocal y_pred
            nonlocal loss
            self.module_.train()
            self.optimizer_.zero_grad()
            y_pred = self.infer(Xi, **fit_params)
            loss = self.get_loss(y_pred, yi, X=Xi, training=True)
            loss.backward()
            self.notify(
            'on_grad_computed',
            named_parameters=list(self.module_.named_parameters())
            )
            return loss

        self.optimizer_.step(train_closure)
        return {
            'loss': loss,
            'y_pred': y_pred,
            }
```

Are there any downsides to this solution? Should I prepare a pull request?

BTW, it could be useful to disable batching, as e.g. LBFGS does not use them.
There is the obvious workaround to set batch size to len(training set), but being able to pass None as batch_size would be clearer, I think

**Comment by ottonemo** [comment]
Thanks for reporting! It is a bit unfortunate that this will make the train step even more complicated than it is now but I don't see another way around it. 

There might be one problem that I can think of when using `nonlocal` variables to communicate the results to the outside: there is no guarantee that the optimizer uses the results of the last called closure, it might use intermediate results for the optimization step. This would introduce a mismatch between returned loss/prediction and actually used loss/prediction between skorch and the optimizer.

For the loss the solution would be simple (`loss = self.optimizer_.step(closure)`) but this doesn't work for the prediction. Maybe we just have to live with it and hope that the optimizer doesn't do something unexpected.

**Comment by benjamin-work** [comment]
Was solved in #247
